### DISCUSSION TITLE
How to Achieve "Spring Hinge joint" with godot jolt

### DISCUSSION THREAD

**Discussion by mgiraud** [godot-jolt/godot-jolt#376] [Questions] [answered]
Godot 4.02
Windows 10
Intel(R) Core(TM) i7-7700K CPU
Nvidia 2080 Super

This is more a call for help or a request feature than an issue ^^. So i'm making a pinball and would like to have a spring hinge joint behaviour for the paddles.
With Godot Physics engine i used some kind of hack :

- When pressing the "paddle up" button, i would set the lower angular limit to the upper angular limit
- When releasing the "paddle up" button, i would set back the lower to it's initial value and set the upper limit to the lower limit value
- The angular bias would simulate the "rotation speed"

I can't do that with godot jolt because:
- The angular bias is not supported by jolt HingeConstraint
- the lower angular limit must be <0 and > -180 and the upper between 0 and 180
- The 6DOF angular spring is not supported by jolt SixDOFConstraint

I've desperatly try all day long to find an alternative
- using animation (terrible idea because of tunneling)
- applying torque impulse, not quite the desired effect and a lot of inertia/torque tweaking...

I'm opened to any alternative that would do the trick

**Comment by mihe** [comment, marked as answer]
I would probably have tried using a torque impulse, like you mention, but I guess you could also try using a motored hinge joint and change the motor's target velocity when you press and release the button. So you'd set the velocity to (for example) +5 when you press the button and -5 when you release the button. You can also tweak the motor's "Max Impulse" property to give it more or less "oompf".

**Reply by mihe** [reply]
It might be worth mentioning that the "Max Impulse" property for hinge motors is a bit of an unfortunate design decision on Godot's part, since you're effectively clamping a value hasn't yet been integrated over time, which means you end up with different results if your timestep ever changes. So if you decide to change the "Physics Ticks per Second" project setting then you will need to tweak these values again. The same applies if you change `Engine.time_scale` for a slowmo effect or something, since that also affects the physics timestep.

6DOF motors don't suffer from this problem since have a force limit instead of an impulse limit. ~~They can however be quite costly in Godot Jolt if you change its properties often, like the target velocity, since they need to be rebuilt entirely every time you do so. It should be fine if you only do it on button press/release though.~~

**Reply by mgiraud** [reply]
Thanks! I will get some sleep and try the 6DOF joint + motor solution tomorrow. I'll edit this post if i'm satisfied with a solution

First Edit : So far i'm almost satisfied with the 6DOF joint + motor. But there is some tunelling (the ball passing through the paddle) and the hit is not consistent
Second edit ; 180 physics FPS + larger paddle is better but tunnelling still happens sometimes,
Third edit : Well so far I have trouble with tunneling, so frustrating

Ok so here what happens : when triggering the paddle it collides the ball a first time, but during the next frame (or more probably the 2nd one) the paddle is rotated (quite a lot to simulate a good kick) and a new collision with the ball may occur if it's not yet far enough fromn the paddle. To fix the issue i've made the collisionshape wayyy thicker so it does not trigger a new collision. I'll try to make a video to illustrate, it works pretty good with this trick

**Comment by mgiraud** [comment]
Here we go:

First a short demo video :
https://github.com/godot-jolt/godot-jolt/assets/11818237/284abd59-afa6-4e83-81de-eed746559cec

I was "lucky" to capture the tunneling effect, here is slow mo of what happened
https://github.com/godot-jolt/godot-jolt/assets/11818237/bc0b9b30-3ad3-4e40-a4c3-9c195339f3c6

What i see :
* The ball does not collide with the paddle when triggering the rotation
* The next frame : the paddle is rotated and passes "through" the ball
* The collision is detected and pushing the ball to the back left

I have to think of a smart solution here, maybe modifying the speculative distance parameter...

And bonus : slow mo of the ball juping over
https://github.com/godot-jolt/godot-jolt/assets/11818237/fa602a02-86a6-4b96-9299-df068cfca4ff

I'll do my best fixing all of this ^^

**Reply by mihe** [reply]
That's pretty cool!

Continuous collision detection, or what Jolt calls "linear cast motion quality", exists to solve this exact issue of tunneling that you're seeing, and some of the examples in Jolt's `Samples` application are much more aggressive than the kinds of speeds you show here, even with triangle-based shapes, so I don't see why it wouldn't work in your application.

If you already have it enabled ("Continuous CD" in Godot) and you're still getting tunneling then it might just be that you need to tweak some of Jolt's physics settings, like the `mLinearCastThreshold` that we talked about in #374. I'll see if I can sneak that into the next release.

**Reply by mgiraud** [reply]
Thanks ! 

CCD is already enabled for the paddles and the ball, so i'll need to tweak the settings i guess.
Or maybe there is some issue with jolt CCD when only angular movement is implied, seems like a long shot though ^^

EDIT : I don't have a decent result when tweaking the `Contact Speculative Distance` or the `Contact Allowed penetration` although the simulation behaves slightly differently each time. Maybe the tweaking of `mLinearCastThreshold` mentionned in #374 will do the trick

**Reply by mgiraud** [reply]
So After a bit of experimentation:
- Gravity = 98, Physics FPS = 120
- CCD is enabled for both paddles ans ball
- The ball has a radius of 0.14m -> 0.28m diameter and has a sphere colldier
- The right paddle is a convex collider generated from a mesh
- The left paddle is a box collider

Here are Godot jolt settings :
![Screenshot 2023-05-23 073952](https://github.com/godot-jolt/godot-jolt/assets/11818237/78d48cb6-3a76-40c2-a355-f1f2d5f478a9)

And i still have **reproducible** tunneling**


https://github.com/godot-jolt/godot-jolt/assets/11818237/117d1073-3fe4-4445-b29b-62ec2a724c0d


https://github.com/godot-jolt/godot-jolt/assets/11818237/1930ec3a-f8ee-490e-bb91-a287177fb67c


https://github.com/godot-jolt/godot-jolt/assets/11818237/4b3bf3e0-5fe0-409a-89a5-f980cf40ea7b


https://github.com/godot-jolt/godot-jolt/assets/11818237/275695a1-5300-47d5-8174-be8f0a6d697a

The only parameter that i have found which fixes the problem is setting the contact speculative much higer (about 0.15m) but of course ghost collisions break the simulation ^^
I'll try to make a MRP today

**Reply by mihe** [reply]
I don't have much experience with using CCD, so forgive my lack of good advice here, but I decided to take a look at how exactly CCD is triggered in Jolt, and when looking at [the code](https://github.com/godot-jolt/jolt/blob/f340da6a18ef83f86ee46b8d7352c79d467c56d9/Jolt/Physics/PhysicsSystem.cpp#L1548) it's pretty clear why this isn't working.

It might very well be that this is how CCD works in every physics engine, but Jolt seems to only check the delta of a body's linear movement when deciding whether to activate CCD or not, and with the paddles not really moving at all (from a linear movement point-of-view) they're presumably never going to have CCD be activated for them. When thinking about it, I guess that makes sense, since you wouldn't really be able to form a straight/linear cast from angular movement.

Having CCD enabled for the ball presumably won't help either, since I guess it's technically the paddle that clips through the ball and not the other way around.

Feel free to create an MRP anyway and we can discuss it further in a separate issue, but I suspect your best shot might be to just bump up the "Physics Ticks per Second" to something even higher.

**Reply by mgiraud** [reply]
Thanks i'll make an issue with the MRP attached, I'll consider this discussion as answered with your last response

**Reply by mihe** [reply]
Considering the original discussion was about rotating the paddles I'll take the liberty of changing the accepted answer. :)
